new/brewblox_service/events.py

```diff
@@ -0,0 +1,308 @@
+        return sub
+
+    async def _listen(self):
+        try:
+            while not self._connection.is_closed:
+                subscription = await self._pending.get()
+
+                # Check whether connection wasn't closed while waiting for queue
+                if not self._connection.is_closed:
+                    await subscription.declare_on_remote(self._connection)
+
+                else:  # pragma: no cover
+                    # Connection closed while retrieving
+                    # Put it back in the queue, we'll declare it after reconnect
+                    self._pending.put_nowait(subscription)
+
+        except RuntimeError as ex:  # pragma: no cover
+            LOGGER.warn(f'Error in {self} {ex}')
+        except CancelledError:
+            LOGGER.info(f'Cancelled {self}')
+
+
+class EventPublisher():
+    def __init__(self, app: Type[web.Application]=None):
+        self._connection = None
+        self._channel = None
+
+        if app:
+            self.setup(app)
+
+    def __str__(self):
+        return f'<{type(self).__name__} {self._connection}>'
+
+    def setup(self, app):
+        app.on_startup.append(self._startup)
+        app.on_cleanup.append(self._cleanup)
+
+    async def _startup(self, app: Type[web.Application]):
+        await self.start(app.loop)
+
+    async def _cleanup(self, app: Type[web.Application]):
+        await self.close()
+
+    async def start(self, loop):
+        def _on_connected(connection):
+            if not connection.is_closed:
+                LOGGER.info(f'Connected {self}')
+
+        self._connection = await aio_pika.connect_robust(loop=loop)
+        self._connection.add_reconnect_callback(_on_connected)
+        _on_connected(self._connection)
+
+    async def close(self):
+        LOGGER.info(f'Closing {self}')
+        if self._connection and not self._connection.is_closed:
+            await self._connection.close()
+
+    async def publish(self,
+                      exchange: str,
+                      routing: str,
+                      message: Union[str, dict]='',
+                      exchange_type: ExchangeType_=ExchangeType.TOPIC):
+
+        # Makes for a more readable error in case of closed connections
+        if not self._connection or self._connection.is_closed:
+            raise ConnectionRefusedError(
+                f'No event bus connection available for {self._connection}')
+
+        # json.dumps() also correctly handles strings
+        data = json.dumps(message).encode()
+
+        if not self._channel:
+            self._channel = await self._connection.channel()
+
+        exchange = await self._channel.declare_exchange(
+            exchange,
+            exchange_type,
+            auto_delete=True)
+
+        # Push it over the line
+        await exchange.publish(
+            aio_pika.Message(data),
+            routing_key=routing
+        )
+
+
+@routes.post('/publish')
```

Comment:
This is temporary for demonstration, right? Should move out of library and into specific plugins?

Reply:
Kept as debugging tool for services.